### ISSUE TITLE
[4.1/master] Gunzip error: -3 (incorrect header check) with ESI

### ISSUE THREAD

**Issue by fgsch** [varnishcache/varnish-cache#1878]
The attached testcase fails under 4.1 and master. Works in 4.0.
Reported by @xinminglai on irc.

**Comment by fgsch** [comment]
[r01878.vtc.txt](https://github.com/varnishcache/varnish-cache/files/172492/r01878.vtc.txt)

**Comment by fgsch** [comment]
For the time being the workaround is to do **"set beresp.do_gunzip = true;"** in `vcl_backend_response{}`.

**Comment by karlvr** [comment]
I have observed the same issue in production on 4.1.1 and 4.1.2. We use ESI, and sometimes ESI within ESI. Our backend servers, Apache 2.2 and Apache 2.4, both have gzip enabled and were delivering gzipped responses for the ESI components.

I have attempted to workaround this while retaining gzipped responses as follows:
```
sub vcl_backend_fetch {
  ...
  unset bereq.http.Accept-Encoding;
  ...
}
```

```
sub vcl_backend_response {
  ...
  if (beresp.http.Content-Type ~ "^text/" || beresp.http.Content-Type ~ "^application/xml" || beresp.http.Content-Type ~ "^application/xhtml" || 
    beresp.http.Content-Type ~ "^application/x-javascript" || beresp.http.Content-Type ~ "^application/javascript") {
    set beresp.do_gzip = true;
  }
  ...
}
```

I make all requests to my backend servers not use the `Accept-Encoding` header, so the backend will never deliver gzip/deflate responses. I then compress responses in Varnish depending upon the `Content-Type` header—effectively mirroring the content type rules that I had setup in my Apache configuration.

Varnish still delivers gzipped responses, so I'm happy. I am currently running this in production and monitoring for errors by running:
```
varnishlog -i Gzip -i ReqUrl -q 'Gzip ~ "error"'
```

I haven't seen any errors in the last 30 minutes (since I've been running with these VCL changes) on Varnish 4.1.2, but the rate at which they occurred was low. I'd expect to see some errors overnight and will report back.
